spinalcordtoolbox/reports/qc2.py

```diff
@@ -0,0 +1,352 @@
+    """
+    if plane not in ['Axial', 'Sagittal']:
+        raise ValueError(f'Invalid plane: {plane!r}')
+
+    logger.info('\n*** Generating Quality Control (QC) html report ***')
+    mod_date = datetime.datetime.now()
+
+    # Not quite following BIDS convention, we derive the value of
+    # dataset, subject, contrast from the input file by splitting it as
+    # {dataset}/{subject}/{contrast}/filename
+    if dataset is None:
+        dataset = path_input.parent.parent.parent.name
+    if subject is None:
+        subject = path_input.parent.parent.name
+    contrast = path_input.parent.name
+    timestamp = mod_date.strftime('%Y_%m_%d_%H%M%S.%f')
+
+    # Make sure the image directory exists
+    path_img = path_qc / dataset / subject / contrast / command / timestamp
+    path_img.mkdir(parents=True, exist_ok=True)
+    path_background_img = path_img / 'background_img.png'
+    path_overlay_img = path_img / 'overlay_img.png'
+
+    # Let the caller generate the image files for the entry
+    yield (path_background_img, path_overlay_img)
+
+    # We lock `index.html` so that we halt any other processes *before*
+    # they have a chance to generate or read any .json files. This ensures
+    # that the last process to write to `index.html` has read in all of the
+    # available .json files, preventing:
+    # https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3701#discussion_r816300380
+    path_index_html = path_qc / 'index.html'
+    with locked_file(path_index_html) as file_index_html:
+
+        # Create a json file for the new QC report entry
+        path_json = path_qc / '_json'
+        path_json.mkdir(parents=True, exist_ok=True)
+        path_result = path_json / f'qc_{timestamp}.json'
+        with path_result.open('w') as file_result:
+            json.dump({
+                'cwd': str(Path.cwd()),
+                'cmdline': cmdline,
+                'command': command,
+                'sct_version': __version__,
+                'dataset': dataset,
+                'subject': subject,
+                'contrast': contrast,
+                'fname_in': path_input.name,
+                'plane': plane,
+                'background_img': str(path_background_img.relative_to(path_qc)),
+                'overlay_img': str(path_overlay_img.relative_to(path_qc)),
+                'moddate': mod_date.strftime("%Y-%m-%d %H:%M:%S"),
+                'qc': '',
+            }, file_result, indent=1)
+
+        # Collect all existing QC report entries
+        json_data = []
+        for path in sorted(path_json.glob('*.json')):
+            with path.open() as file:
+                json_data.append(json.load(file))
+
+        # Insert the QC report entries into index.html
+        path_assets = importlib.resources.files(spinalcordtoolbox.reports) / 'assets'
+        template = string.Template((path_assets / 'index.html').read_text(encoding='utf-8'))
```

Comment:
On second thought, I guess `string.Template` is fine. It's just that shell-style substitution tends to be so full of subtle pitfalls that it immediately raises my hackles. But the rules for `string.Template` are simple enough that I don't think it's a problem.

(It didn't help that the `safe_substitute` method is what I would really consider to be the **unsafe** version of the `substitute` method, [as acknowledged in the docs](https://docs.python.org/3/library/string.html#string.Template.safe_substitute). The kind of thinking behind that name is exactly what leads to subtle pitfalls!)

Still, our usage is so simple that it could be replaced with a call to the builtin `str.replace` method.